src/app/reporting/ReportSchedulerImpl.cpp

```diff
@@ -0,0 +1,194 @@
+        Milliseconds32(node->GetMaxInterval().count() - System::SystemClock().GetMonotonicTimestamp().count());
+    ScheduleReport(newTimeout, apReadHandler);
+}
+
+/// @brief When a ReadHandler is removed, unregister it, which will cancel any scheduled report
+void ReportSchedulerImpl::OnReadHandlerRemoved(ReadHandler * aReadHandler)
+{
+    UnregisterReadHandler(aReadHandler);
+}
+
+CHIP_ERROR ReportSchedulerImpl::RegisterReadHandler(ReadHandler * aReadHandler)
+{
+    ReadHandlerNode * newNode = FindReadHandlerNode(aReadHandler);
+    // If the handler is already registered, no need to register it again
+    VerifyOrReturnValue(nullptr == newNode, CHIP_NO_ERROR);
+    newNode = mNodesPool.CreateObject(aReadHandler);
+    mReadHandlerList.PushBack(newNode);
+
+    ChipLogProgress(DataManagement,
+                    "Registered ReadHandler that will schedule a report between system Timestamp: %" PRIu64
+                    " and system Timestamp %" PRIu64 ".",
+                    newNode->GetMinInterval().count(), newNode->GetMaxInterval().count());
+
+    Timestamp now = System::SystemClock().GetMonotonicTimestamp();
+    // If the handler is reportable, schedule a report for the min interval, otherwise schdule a report for the max interval
+    if ((newNode->IsReportableNow()))
+    {
+        // If the handler is reportable now, just schedule a report immediately
+        InteractionModelEngine::GetInstance()->GetReportingEngine().ScheduleRun();
+    }
+    else if (IsReadHandlerReportable(aReadHandler) && (newNode->GetMinInterval() > now))
+    {
+        Milliseconds32 newTimeout = Milliseconds32(newNode->GetMinInterval().count() - now.count());
+        // If the handler is reportable now, but the min interval is not elapsed, schedule a report for the min interval
+        ReturnErrorOnFailure(ScheduleReport(newTimeout, aReadHandler));
+    }
+    else
+    {
+        Milliseconds32 newTimeout = Milliseconds32(newNode->GetMaxInterval().count() - now.count());
+        // If the handler is not reportable now, schedule a report for the max interval
+        ReturnErrorOnFailure(ScheduleReport(newTimeout, aReadHandler));
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ReportSchedulerImpl::ScheduleReport(Timeout timeout, ReadHandler * aReadHandler)
+{
+    // Verify the handler is still registered
+    ReadHandlerNode * node = FindReadHandlerNode(aReadHandler);
+    VerifyOrReturnError(nullptr != node, CHIP_ERROR_NOT_FOUND);
+
+    // Cancel Report if it is currently scheduled
+    InteractionModelEngine::GetInstance()->GetExchangeManager()->GetSessionManager()->SystemLayer()->CancelTimer(
+        ReportTimerCallback, node);
+
+    if (!IsChunkedReport(aReadHandler))
+    {
+        // Schedule Report
+        ReturnErrorOnFailure(
+            InteractionModelEngine::GetInstance()->GetExchangeManager()->GetSessionManager()->SystemLayer()->StartTimer(
+                timeout, ReportTimerCallback, node));
+    }
```

Comment:
Suggest making a "ReportScheduler::TimerDelegate" inner interface that is registered with the ReportScheduler. This would handle the StartTimer and CancelTimer hook-up. This way, you can do mock checks of timers during unit tests, and you don't need to tie yourself 100% down to directly calling 4 layers of getters + direct dep on the SystemLayer. It could also handle the adaptation between SystemLayer Timer callback and hitting the actual ReportTimerCallback.

Just a thought. Can be done as follow-up if it's too much work.

Reply:
I just added a TimerDelegate class in the ReportScheduler, let me know if it matches what you had in mind and I'll complete the implementation